Allow Emoji Characters in the URL

AntonNguyen opened the pull request:
Emoji characters are a valid in a URL, so the validator should be updated to allow this edge case to pass.

Fun Emoji/Browser facts
1. Browsers utilize something called `punycoder` to convert something like `🍕` to `xn--vi8h`
2. At the moment, the only ccTLD that supports emojis in the domain name, is `.ws`!

refs #8368

AntonNguyen committed:
Allow utf8mb4 characters in the URL
refs #8368

burzum commented:
Looks good. But could we add a `@link` doc block to some RFC or spec for the emoji stuff? I haven't been aware that there is something like that and also not that you're allowed to use them in URLs.

markstory commented on src/Validation/Validation.php:
```diff
@@ -908,7 +908,9 @@ public static function range($check, $lower = null, $upper = null)
         static::_populateIp();
-        $alpha = '0-9\p{L}\p{N}';
+
+        $emoji = '\x{1F300}-\x{1F6FF}';
```
Should we go higher than U+1F6FF? The [emoji table on wikipedia](https://en.wikipedia.org/wiki/Emoji) has 🧀 at U+1F9C0. The points above 1F6FF are a bit sparse but there are a few good ones up there.

AntonNguyen replied:
Absolutely!

AntonNguyen committed:
Widened the set of allowable emojis.
refs #8368

AntonNguyen committed:
Added some documentation on what makes a valid URI.

From the docs, it says,
“A URI is composed from a limited set of characters consisting of
   digits, letters, and a few graphic symbols”

refs #8368

ADmad commented on src/Validation/Validation.php:
```diff
@@ -904,11 +904,14 @@ public static function range($check, $lower = null, $upper = null)
-        $alpha = '0-9\p{L}\p{N}';
+
+        $emoji = '\x{1F190}-\x{1F9EF}';
+        $alpha = '0-9(\p{L}\p{N}' . $emoji;
```
Is the addition of `(` in the regex intentional? It wasn't there earlier.

AntonNguyen replied:
Nice catch!

markstory replied:
I can fix that before merging.

saeideng commented:
maybe `urlWithEmoji()`  instead of `url()`

markstory commented:
@saeideng Why a separate method?

saeideng commented:
maybe someone unlike the emoji in Url

markstory commented:
Someone not liking emoji in urls doesn't make those URLs less valid though.

inoas commented:
It would be great if the validator allowed validation of:
- asciiOnlyUrl
- utf8UrlNoEmoji
- utf8Url

... or similar.

markstory commented:
@inoas I don't think we need to meet all those use cases in the validators that CakePHP provides. If people need to support a subset of valid URLs they can create their own patterns.

inoas commented:
Well it may be a valid URL but that doesn't mean it is a good URL and the validator before allowing emojis > the validator after allowing them (also it is not BC technically).

dereuromark commented:
I am with @markstory on this one:
If the current http specs allow for these URLs, and those are technically valid/correct, then this warrants a fix in a patch release.

If you don't like them, feel free to either use your own validation rule for URLs or simply add a validation rule that sniffs for the '\x{1F190}-\x{1F9EF}' 4 byte range and disallows their use. I don't think this is needed or justified, however, in the majority of use cases.

markstory commented:
Thanks!
